### PR TITLE
#67 chore: improve development experience for local setup

### PR DESCRIPTION
**Pull request by ternakkode**
## Changes : 

1. Integratation with [golang-migrate](https://github.com/golang-migrate/migrate) library to sync the database schema
2. All docker compose is stored in `infra/local` path
3. Any new configuration to setup the dependency for running the layer8server should added to `server/cmd/setup/setup.go`

### Covered setups : 
1. Initialize PostgreSQL Database and run database migration.
2. Intialize InfluxDB Database and create authentication token.
4. Initialize Telegraf container.

## How to run : 

1. `make setup_local_dependency` to run all required container
5. `make run_layer8server_local` to run layer8 server
6. `make setup_and_run` to both setup and run layer8 

Refer to [Contributor Guideline](https://github.com/globe-and-citizen/layer8/wiki/Contributor-Guideline:-Setting-Up-Local-Environment) to understand how to contribute a new container for local development

### CODE REVIEW

**Review by stravid87 — CHANGES_REQUESTED** [review]
Awesome work! This is huge progress. 
Seems like the docker images run. And it's nicely organized.
You've got exactly the right idea with the simple make commands. 
However, on running, looks like I'm still getting an error to register a client:
Navigating to `http://localhost:5001/client-register-page` I get
"open assets-v1/templates/src/pages/client_portal/register.html: The system cannot find the path specified."

**Comment by stravid87** [comment]
Great work! In my mind this was a 2 week task. Beautiful. My only feedback:
1) do a manual testing checklist
2) double check the stats... maybe this needs another step?



## Test Cases Checklist

### SECTION 1: RESOURCE SERVER (http://localhost:5001/)
- [ ] http://localhost:5001/ home page loads
- [ ] "LoginAsUser" button takes you to the user login page
- [ ] Log in using "tester" and "12341234" should succeed
- [ ] Updating the display name should work
- [ ] Log out and Register a new user and try to log in with that new user
- [ ] "LoginAsClient" button takes you to the client login page
- [ ] Log in using "layer8" and "12341234" should work
- [ ] You will be able to see your UID and Secret in profile
- [ ] Log out and Register a new client and log in using that should work

### SECTION 2: WE'VE GOT POEMS (http://localhost:5173/)
- [ ] Log in with 'tester' / '1234'
- [ ] Log in anonymously with Layer8
- [ ] Clicking "Get Next Poem" loads different poem correctly x 3
- [ ] Clicking "Logout" takes you to the login screen
- [ ] Clicking "Register" takes you to the registration page
- [ ] Registering with a username, password, and profile image is successful
- [ ] Logging in with the new username / password succeeds
- [ ] Logging in with Layer8 opens the pop-up
- [ ] Logging in with the newly registered credentials from Section 1 succeeds
- [ ] User chooses to share their new "Username" & "Country" from the Layer8 Resource Server
- [ ] Clicking "Get Next Poem" loads different poem correctly x 3
- [ ] Clicking "Logout" takes you to the login page
- [ ] Log in with 'tester' / '1234', and then with the credentials from Section 1 succeeds
- [ ] This time, DO NOT share "display name" or "country"
- [ ] Clicking "Get Next Poem" loads different poem correctly x 3

### SECTION 3: IMSHARER (http://localhost:5174/)
- [ ] Main page loads
- [ ] Upload of image works
- [ ] Reload leads to instant reload (demonstrating proper caching)
- [ ] Clicking the newly loaded image shows it in a lightbox
- [ ] Log in using tester and 12341234 should succeed

**Comment by ternakkode** [comment]
> Great work! In my mind this was a 2 week task. Beautiful. My only feedback:
> 
> 1. do a manual testing checklist
> 2. double check the stats... maybe this needs another step?
> 
> ## Test Cases Checklist
> ### SECTION 1: RESOURCE SERVER (http://localhost:5001/)
> * [x]  http://localhost:5001/ home page loads
> * [x]  "LoginAsUser" button takes you to the user login page
> * [x]  Log in using "tester" and "12341234" should succeed
> * [x]  Updating the display name should work
> * [x]  Log out and Register a new user and try to log in with that new user
> * [x]  "LoginAsClient" button takes you to the client login page
> * [x]  Log in using "layer8" and "12341234" should work
> * [x]  You will be able to see your UID and Secret in profile
> * [x]  Log out and Register a new client and log in using that should work
> 
> ### SECTION 2: WE'VE GOT POEMS (http://localhost:5173/)
> * [x]  Log in with 'tester' / '1234'
> * [x]  Log in anonymously with Layer8
> * [x]  Clicking "Get Next Poem" loads different poem correctly x 3
> * [x]  Clicking "Logout" takes you to the login screen
> * [x]  Clicking "Register" takes you to the registration page
> * [x]  Registering with a username, password, and profile image is successful
> * [x]  Logging in with the new username / password succeeds
> * [x]  Logging in with Layer8 opens the pop-up
> * [x]  Logging in with the newly registered credentials from Section 1 succeeds
> * [x]  User chooses to share their new "Username" & "Country" from the Layer8 Resource Server
> * [x]  Clicking "Get Next Poem" loads different poem correctly x 3
> * [x]  Clicking "Logout" takes you to the login page
> * [x]  Log in with 'tester' / '1234', and then with the credentials from Section 1 succeeds
> * [x]  This time, DO NOT share "display name" or "country"
> * [x]  Clicking "Get Next Poem" loads different poem correctly x 3
> 
> ### SECTION 3: IMSHARER (http://localhost:5174/)
> * [x]  Main page loads
> * [x]  Upload of image works
> * [x]  Reload leads to instant reload (demonstrating proper caching)
> * [x]  Clicking the newly loaded image shows it in a lightbox
> * [ ]  Log in using tester and 12341234 should succeed

i had some problem with the test case, especially with the imgsharer part : 
- i dont find any login function in imgsharer
- imsharer frontend is by default running in port 5173 not 5174
- imsharer backend by default is running in port 6001 so we need to register a client with port 6001 first before we can use imsharer

i propose to update the guideline and change the imsharer backend port to 8000 (the same as wgp backend) for the sake of simplicity

**Comment by DeeStarks** [comment]
> @stravid87
> 
> i had some problem with the test case, especially with the imsharer part, here are the issues :
> 
> * the interceptor sometimes giving panic error after i logged in as layer8, not sure how to reproduce it
> * i dont find any login function in imsharer
> * imsharer frontend is by default running in port 5173 not 5174
> * imsharer backend by default is running in port 6001 so we need to register a client with port 6001 first before we can use imsharer
> 
> i propose to update the guideline and change the imsharer backend port to 8000 (the same as wgp backend) for the sake of simplicity

@ternakkode 

- Could you provide more details on the error in the interceptor? A screenshot for example
- Imsharer was built as a provider with no dependency on layer8's authentication mechanism but uses only the tunnel, which is a good test case for clients that want to build simple websites that don't need users to register or login but still want the anonymization feature of the layer8 tunnel (e.g. users IP masking)
- It is a vue app, vue automatically assigns the next available port number for new apps when the default port is already allocated. It'll run on 5174 when you have wgp already running on 5173 for example. However, I think the point is to test that it works - the port shouldn't be a blocker
- I might have to disagree with the proposal because the reason for mocks is to allow us to simulate various scenarios such as edge cases and error conditions. Imsharer represents a client, and rather than enforcing it to use a specific port, we should allow it to choose its preferred port. What I would rather is to register port 6001 as a client through the layer8 dashboard/portal and check that it works as expected when we use a port other than 8000

File: migrations/000001_init.up.sql [hunk 1]
```diff
@@ -0,0 +1,69 @@
+CREATE TABLE clients (
```
Comment (DeeStarks) [inline migrations/000001_init.up.sql:1]: What do you think about [Prisma](https://www.prisma.io/docs)? I find it a great option and we wouldn't need to write raw SQL queries/schemas. The [Go client](https://goprisma.org/) makes the work pretty easy

CC @stravid87 @huzaifamk

File: sp_mocks/imsharer/backend/index.js [hunk 2]
```diff
@@ -4,7 +4,7 @@ const fs = require('fs');
 const layer8 = require('layer8_middleware');
 
 const app = express();
-const port = 6001;
+const port = 8000;
```
Comment (DeeStarks) [inline sp_mocks/imsharer/backend/index.js:7]: I would rather we leave this as 6001 and instead test the capability of the proxy on ports like this other than 8000

File: sp_mocks/imsharer/frontend/src/App.vue [hunk 3]
```diff
@@ -6,7 +6,7 @@ import emitter from '@/plugins/mitt';
 
 onMounted(async () => {
   await layer8.initEncryptedTunnel({
-    providers: ["http://localhost:6001"],
+    providers: ["http://localhost:8000"],
```
Comment (DeeStarks) [inline sp_mocks/imsharer/frontend/src/App.vue:9]: We should leave this as 6001 along with the others and register it as a new client instead

Comment (ternakkode) [reply sp_mocks/imsharer/frontend/src/App.vue:9]: sure, i change it so it follow the manual test case, we can update the test case then

File: migrations/000001_init.up.sql (hunk 1 above)
Comment (ternakkode) [reply migrations/000001_init.up.sql:1]: We already have [GORM](https://gorm.io/index.html) for query builder / ORM.

Comment (ternakkode) [reply migrations/000001_init.up.sql:1]: Regarding the schema migration, I'm more familiar with golang-migrate and writing raw queries rather than using an ORM to migrate the database structure from the struct definition. I'm not sure how the library generates the database schema so i feel that it more convinient to write raw DDL.

However, I'm not against using an ORM if everyone feels it's more beneficial for managing the database schema.

**Review by huzaifamk — APPROVED** [review]
The two make commands works well!

**Review by DeeStarks — APPROVED** [review]
